pkg/v1/mutate/oci.go

```diff
@@ -0,0 +1,162 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mutate
+
+import (
+	"fmt"
+	"io"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/match"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+// Drops docker specific properties
+// See: https://github.com/opencontainers/image-spec/blob/main/config.md
+func toOCIV1Config(config v1.Config) v1.Config {
+	return v1.Config{
+		User:         config.User,
+		ExposedPorts: config.ExposedPorts,
+		Env:          config.Env,
+		Entrypoint:   config.Entrypoint,
+		Cmd:          config.Cmd,
+		Volumes:      config.Volumes,
+		WorkingDir:   config.WorkingDir,
+		Labels:       config.Labels,
+		StopSignal:   config.StopSignal,
+	}
+}
+
+func toOCIV1ConfigFile(cf *v1.ConfigFile) *v1.ConfigFile {
+	return &v1.ConfigFile{
+		Created:      cf.Created,
+		Author:       cf.Author,
+		Architecture: cf.Architecture,
+		OS:           cf.OS,
+		OSVersion:    cf.OSVersion,
+		History:      cf.History,
+		RootFS:       cf.RootFS,
+		Config:       toOCIV1Config(cf.Config),
+	}
+}
+
+// OCIImage mutates the provided v1.Image to be OCI compliant v1.Image
+// Check image-spec to see which properties are ported and which are dropped.
+// https://github.com/opencontainers/image-spec/blob/main/config.md
+func OCIImage(base v1.Image) (v1.Image, error) {
+	// Get original manifest
+	m, err := base.Manifest()
+	if err != nil {
+		return nil, err
+	}
+	// Convert config
+	cfg, err := base.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+	cfg = toOCIV1ConfigFile(cfg)
+
+	layers, err := base.Layers()
+	if err != nil {
+		return nil, err
+	}
+
+	newLayers := []v1.Layer{}
+
+	for _, layer := range layers {
+		mediaType, err := layer.MediaType()
+		if err != nil {
+			return nil, err
+		}
+		switch mediaType {
+		case types.DockerLayer:
+			reader, err := layer.Compressed()
+			if err != nil {
+				return nil, fmt.Errorf("getting layer: %w", err)
+			}
+			layer, err = tarball.LayerFromOpener(func() (io.ReadCloser, error) { return reader, nil }, tarball.WithMediaType(types.OCILayer))
+			if err != nil {
+				return nil, fmt.Errorf("building layer: %w", err)
+			}
+		case types.DockerUncompressedLayer:
+			reader, err := layer.Uncompressed()
+			if err != nil {
+				return nil, fmt.Errorf("getting layer: %w", err)
+			}
+			layer, err = tarball.LayerFromOpener(func() (io.ReadCloser, error) { return reader, nil }, tarball.WithMediaType(types.OCIUncompressedLayer))
+			if err != nil {
+				return nil, fmt.Errorf("building layer: %w", err)
+			}
+		}
+		newLayers = append(newLayers, layer)
+	}
```

Comment:
I suspect there is a better way to do this, this is rather expensive.

Reply:
I'd expect that but i wasn't able to find it.

Reply:
ping @jonjohnsonjr do you have suggestions to make this less expensive? This PR seems stuck.

Reply:
This is annoying because of how `partial.Descriptor` works, but I would probaby create a new type, something like:

```go
type layerWithMediaType struct {
  v1.Layer
  mt types.MediaType
}

func (l *layerWithMediaType) Descriptor() (*v1.Descriptor, error) {
  // call partial.Descriptor on wrapped layer and update MediaType
}

func (l *layerWithMediaType) MediaType() (types.MediaType, error) {
  return l.mt, nil
}
```